### PR TITLE
C++: Use pyrameterized modules for TaintTracking 

### PR DESCRIPTION
**Pull request by jbj**
This is a minimal PR to demonstrate how taint tracking can be pyrameterized such that each copy of the taint-tracking configuration class is byte-for-byte identical. It also switches the C++ interface from using `TaintTracking::Configuration2` to using `TaintTracking2::Configuration` like C# does it.

The commits should make sense individually: the first commit rearranges the data flow library, and the second commit makes the actual change to C++ taint tracking.

I didn't go as far as unifying the taint-tracking configuration classes across languages. I can follow up with language-specific PRs for that if we merge this PR.

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/dataflow/TaintTracking2.qll [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+ * Provides classes for performing local (intra-procedural) and
+ * global (inter-procedural) taint-tracking analyses.
+ *
+ * We define _taint propagation_ informally to mean that a substantial part of
+ * the information from the source is preserved at the sink. For example, taint
+ * propagates from `x` to `x + 100`, but it does not propagate from `x` to `x >
+ * 100` since we consider a single bit of information to be too little.
+ */
+module TaintTracking2 {
+  import semmle.code.cpp.dataflow.internal.TaintTrackingUtil
```
Comment (hvitved) [inline cpp/ql/src/semmle/code/cpp/dataflow/TaintTracking2.qll]: Shouldn't this import happen in `dataflow2/TaintTrackingImpl.qll` (and same for `TaintTracking.qll`)?

Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/dataflow/TaintTracking2.qll]: It probably should. That would be consistent with how we do it for `DataFlow`. I'll change that.

**Comment by jbj** [comment]
I simplified this PR vastly after talking to @aschackmull. I dropped the commit that changed the data-flow libraries, so this PR now only changes the C/C++ taint tracking library. That means fewer code changes and fewer confusing indirections.

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/tainttracking2/TaintTrackingParameter.qll [hunk 2]
```diff
@@ -0,0 +1,2 @@
+import semmle.code.cpp.dataflow.DataFlow as Private
```
Comment (hvitved) [inline cpp/ql/src/semmle/code/cpp/dataflow/internal/tainttracking2/TaintTrackingParameter.qll]: Should be `dataflow2`, right?

Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/dataflow/internal/tainttracking2/TaintTrackingParameter.qll]: Thanks for spotting!

Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/dataflow/internal/tainttracking2/TaintTrackingParameter.qll]: Fixed.

**Comment by jbj** [comment]
Fixing the `DataFlow2` bug that Tom found revealed that `TaintTrackingParameter.qll` has to do a little more work: it has to rename `DataFlow2` to `DataFlow` in the scope of `tainttracking2/TaintTrackingImpl.qll`. I added that in  commit no. 3.

@aschackmull, are you happy enough with these changes that I should apply the same restructuring to Java (and make the taint tracking implementation files identical across languages)?

**Comment by aschackmull** [comment]
> @aschackmull, are you happy enough with these changes that I should apply the same restructuring to Java (and make the taint tracking implementation files identical across languages)?

Yes.

File: java/ql/src/semmle/code/java/dataflow/TaintTracking.qll [hunk 3]
```diff
@@ -1,804 +1,24 @@
 /**
  * Provides classes for performing local (intra-procedural) and
  * global (inter-procedural) taint-tracking analyses.
+ *
+ * We define _taint propagation_ informally to mean that a substantial part of
+ * the information from the source is preserved at the sink. For example, taint
+ * propagates from `x` to `x + 100`, but it does not propagate from `x` to `x >
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/dataflow/TaintTracking.qll]: This does not hold for java.

File: java/ql/src/semmle/code/java/dataflow/TaintTracking2.qll [hunk 4]
```diff
@@ -0,0 +1,12 @@
+/**
+ * Provides classes for performing local (intra-procedural) and
+ * global (inter-procedural) taint-tracking analyses.
+ *
+ * We define _taint propagation_ informally to mean that a substantial part of
+ * the information from the source is preserved at the sink. For example, taint
+ * propagates from `x` to `x + 100`, but it does not propagate from `x` to `x >
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/dataflow/TaintTracking2.qll]: Same.

**Comment by jbj** [comment]
I now also made the change for C# and added change notes. I rebased the PR to avoid a change note conflict.

File: csharp/ql/src/semmle/code/csharp/dataflow/internal/tainttracking1/TaintTrackingParameter.qll [hunk 5]
```diff
@@ -0,0 +1,5 @@
+import semmle.code.csharp.dataflow.internal.TaintTrackingUtil as Public
+
+module Private {
+  import semmle.code.csharp.dataflow.DataFlow::DataFlow as DataFlow
+}
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/dataflow/internal/tainttracking1/TaintTrackingParameter.qll]: I would prefer
```
import semmle.code.csharp.dataflow.internal.TaintTrackingPublic as Public

module Private {
  import semmle.code.csharp.dataflow.DataFlow::DataFlow as DataFlow
  import semmle.code.csharp.dataflow.internal.TaintTrackingPrivate
}
```
and get rid of `TaintTrackingUtil.qll`

Comment (jbj) [reply csharp/ql/src/semmle/code/csharp/dataflow/internal/tainttracking1/TaintTrackingParameter.qll]: Done

**Pull request merged**